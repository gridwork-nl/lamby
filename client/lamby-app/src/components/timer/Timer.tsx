import { useState, useEffect, useContext } from 'react';
import { AppStateContext } from '../../context/AppContext';


interface ITimeLeft {
  days: number;
  hours: number;
  minutes: number;
  seconds: number;
}

const Timer: React.FC = () => {
  // Context Hook
  const { data } = useContext(AppStateContext);
  const match = data.match;
  console.log(match);

  // Function to get the values for the timer
  const countDownTimer = (): ITimeLeft => {
    const matchTime: number = new Date(match.date).getTime();
    const now: number = new Date().getTime();
    const diff: number = matchTime - now;
    let timeLeft = {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
    };
    if (diff > 0) {
      timeLeft = {
        days: Math.floor(diff / (1000 * 60 * 60 * 24)),
        hours: Math.floor((diff / (1000 * 60 * 60)) % 24),
        minutes: Math.floor((diff / 1000 / 60) % 60),
        seconds: Math.floor((diff / 1000) % 60),
      };
    }
    return timeLeft;
  };

  // update the time left values to use in de
  const [timeLeft, setTimeLeft] = useState<ITimeLeft>(countDownTimer());

  useEffect(() => {
    let isMounted = true;
    setTimeout(() => {
      if (isMounted) setTimeLeft(countDownTimer());
    }, 1000);
    return () => {
      isMounted = false;
    };
  });
  return (
    <div className='timer'>
      <span data-testid='day'>
        {timeLeft.days}
        {timeLeft.days > 1 ? ' DAYS ' : ' DAY '}
      </span>
      <span data-testid='hour'>
        {timeLeft.hours}
        {timeLeft.hours > 1 ? ' HOURS ' : ' HOUR '}
      </span>
      <span data-testid='mins'>
        {timeLeft.minutes}
        {timeLeft.minutes > 1 ? ' MINS ' : ' MIN '}
      </span>
      <span data-testid='sec'>
        {timeLeft.seconds}
        {timeLeft.seconds > 1 ? ' SECS ' : ' SEC '}
      </span>
    </div>
  );
};

export default Timer;
