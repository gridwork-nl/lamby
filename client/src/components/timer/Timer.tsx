import { useState, useEffect, useContext } from 'react'
import { ITimeLeft } from '../../interfaces/interfaces'
import { AppStateContext } from '../../context/AppContext'

const Timer: React.FC = () => {
// Context Hook
  const { data } = useContext(AppStateContext)
  const match = data.match

  // Function to get the values for the timer
  const countDownTimer = (): ITimeLeft => {
    const matchTime: number = new Date(match.date).getTime()
    const now: number = new Date().getTime()
    const diff: number = matchTime - now
    let timeLeft = {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0
    }
    if (diff > 0) {
      timeLeft = {
        days: Math.floor(diff / (1000 * 60 * 60 * 24)),
        hours: Math.floor((diff / (1000 * 60 * 60)) % 24),
        minutes: Math.floor((diff / 1000 / 60) % 60),
        seconds: Math.floor((diff / 1000) % 60)
      }
    }
    return timeLeft
  }

  // update the time left values to use in de
  const [timeLeft, setTimeLeft] = useState<ITimeLeft>(countDownTimer())

  useEffect(() => {
    let isMounted = true
    setTimeout(() => {
      if (isMounted) setTimeLeft(countDownTimer())
    }, 1000)
    return () => { isMounted = false }
  })

  return (
    <div className="timer">
      <span>
        {timeLeft.days}
        {timeLeft.days > 1 ? ' DAYS ' : ' DAY '}
      </span>
      <span>
        {timeLeft.hours}
        {timeLeft.hours > 1 ? ' HOURS ' : ' HOUR '}
      </span>
      <span>
        {timeLeft.minutes}
        {timeLeft.minutes > 1 ? ' MINS ' : ' MIN '}
      </span>
      <span>
        {timeLeft.seconds}
        {timeLeft.seconds > 1 ? ' SECS ' : ' SEC '}
      </span>
    </div>
  )
}

export default Timer
