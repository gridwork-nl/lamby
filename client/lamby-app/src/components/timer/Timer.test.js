import ReactDom from 'react-dom';
import React from 'react';
import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ShallowRenderer from 'react-test-renderer/shallow';
import Timer from './Timer';

const timerMock = {
  data: {
    match: {
    id: 0,
    homeTeam: 'barca',
    awayTeam: 'madrid',
    formation: '4-3-3',
    venue: 'Old Trafford',
    date:"2021-04-14T15:00:00.000Z" ,
<<<<<<< HEAD
   
    }
   
  
=======
    }
>>>>>>> testing-back-end
  }
}



it('should log a user in', () => {
const AppStateContext = React.createContext(timerMock);
<<<<<<< HEAD
    const { getByText } = render(
=======
 render(
>>>>>>> testing-back-end
    < MyContext.Provider; value={data}>
    <Timer />
  </ MyContext.Provider>
  );
<<<<<<< HEAD
    
=======
>>>>>>> testing-back-end
 })


afterEach(cleanup);

it('it renders Time component', () => {
  const div = document.createElement('div');
  ReactDom.render(<Timer></Timer>, div);
  ReactDom.unmountComponentAtNode(div);
});

it('it render the time left', () => {
  const { getByTestId } = render(<Timer />);
  const day = getByTestId('day');
  expect(day).toBeInTheDocument();
  const hour = getByTestId('hour');
  expect(hour).toBeInTheDocument();
  const mins = getByTestId('mins');
  expect(mins).toBeInTheDocument();
  const sec = getByTestId('sec');
  expect(sec).toBeInTheDocument();
});

test('mock hook', () => {
  timerContextMock.mockReturnValue(AppStateContext);
  const { getByTestId } = new ShallowRenderer().render(<Timer />);
  const time = getByTestId('day');
  console.log(time.innerHTML);
  console.log();
});
