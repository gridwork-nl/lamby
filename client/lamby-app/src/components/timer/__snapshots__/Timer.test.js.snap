// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly Home-Page 1`] = `
<div
  className="timer"
>
  <span
    data-testid="day"
  >
    0
     DAY 
  </span>
  <span
    data-testid="hour"
  >
    0
     HOUR 
  </span>
  <span
    data-testid="mins"
  >
    0
     MIN 
  </span>
  <span
    data-testid="sec"
  >
    0
     SEC 
  </span>
</div>
`;
